session/tlv.go

```diff
@@ -150,5 +147,46 @@ func DeserializeSession(r io.Reader) (*Session, error) {
 		)
 	}
 
+	if t, ok := parsedTypes[typeMacaroonRecipe]; ok && t == nil {
+		recipe, err := macaroonPermsFromString(string(macRecipe))
+		if err != nil {
+			return nil, err
+		}
+
+		session.MacaroonRecipe = recipe
+	}
+
 	return session, nil
 }
+
+func macaroonPermsToString(perms []bakery.Op) string {
+	var macRecipe string
+	for _, op := range perms {
+		if macRecipe != "" {
+			macRecipe += ";"
+		}
+
+		macRecipe += fmt.Sprintf("%s:%s", op.Entity, op.Action)
+	}
+
+	return macRecipe
+}
+
+func macaroonPermsFromString(macRecipe string) ([]bakery.Op, error) {
+	perms := strings.Split(macRecipe, ";")
+	recipe := make([]bakery.Op, 0, len(perms))
+	for _, permStr := range perms {
+		perm := strings.Split(permStr, ":")
+		if len(perm) != 2 {
+			return nil, errors.New("invalid macaroon " +
+				"recipe encoding")
+		}
+
+		recipe = append(recipe, bakery.Op{
+			Entity: perm[0],
+			Action: perm[1],
```

Comment:
gonna replace all this. just found tlv.MakeDynamicRecord. gonna use that instead. also need to add caveates